File: src/main/java/komu/blunt/parser/DataTypeParser.java
package komu.blunt.parser;

import com.google.common.collect.ImmutableList;
import komu.blunt.ast.AST;
import komu.blunt.ast.ASTDataDefinition;
import komu.blunt.types.*;

import java.util.ArrayList;
import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;
import static komu.blunt.parser.TokenType.*;
import static komu.blunt.types.Qualified.quantify;
import static komu.blunt.types.Type.functionType;
import static komu.blunt.types.Type.genericType;

final class DataTypeParser {
    
    private final Lexer lexer;
    private final TypeParser typeParser;

    DataTypeParser(Lexer lexer, TypeParser typeParser) {
        this.lexer = checkNotNull(lexer);
        this.typeParser = checkNotNull(typeParser);
    }

    // data <type> <var>* = <constructor>+
    ASTDataDefinition parseDataDefinition() {
        lexer.expectIndentStartToken(DATA);

        DataTypeBuilder builder = new DataTypeBuilder(lexer.readTokenValue(TYPE_OR_CTOR_NAME));
        
        while (!lexer.readMatchingToken(ASSIGN))
            builder.addVariable(typeParser.parseTypeVariable());

        do {
            parseConstructor(builder);
        } while (lexer.readMatchingToken(OR));

        if (lexer.readMatchingToken(DERIVING)) {
            lexer.expectToken(LPAREN);
            do {
                builder.addAutomaticallyDerivedClass(lexer.readTokenValue(TYPE_OR_CTOR_NAME));
            } while (lexer.readMatchingToken(COMMA));
            lexer.expectToken(RPAREN);
        }

        lexer.expectToken(END);

        return builder.build();
    }

    private void parseConstructor(DataTypeBuilder builder) {
        String constructorName = lexer.readTokenValue(TYPE_OR_CTOR_NAME);

        List<Type> args = new ArrayList<Type>();
        while (!lexer.nextTokenIs(OR) && !lexer.nextTokenIs(END) && !lexer.nextTokenIs(DERIVING))
            args.add(typeParser.parseTypePrimitive());

        builder.addConstructor(constructorName, args);
    }

    private static final class DataTypeBuilder {

        private final String typeName;
        private final List<TypeVariable> vars = new ArrayList<TypeVariable>();
        private final ImmutableList.Builder<ConstructorDefinition> constructors = ImmutableList.builder();
        private final ImmutableList.Builder<String> derivedClasses = ImmutableList.builder();
        private int constructorIndex = 0;

        public DataTypeBuilder(String typeName) {
            this.typeName = checkNotNull(typeName);
        }
        
        public void addVariable(TypeVariable variable) {
            vars.add(checkNotNull(variable));
        }

        public void addConstructor(String constructorName, List<Type> args) {
            Scheme scheme = quantify(vars, new Qualified<Type>(functionType(args, getType())));
            constructors.add(new ConstructorDefinition(constructorIndex++, constructorName, scheme, args.size()));
        }

        private Type getType() {
            return genericType(typeName, vars);
        }

        public void addAutomaticallyDerivedClass(String className) {
            derivedClasses.add(className);   
        }
        
        public ASTDataDefinition build() {
            return AST.data(typeName, getType(), constructors.build(), derivedClasses.build());
        }
    }
}

File: src/main/kotlin/komu/blunt/parser/DataTypeParser.kt
package komu.blunt.parser

import komu.blunt.ast.ASTDefinition
import komu.blunt.parser.TokenType.Companion.ASSIGN
import komu.blunt.parser.TokenType.Companion.COMMA
import komu.blunt.parser.TokenType.Companion.DATA
import komu.blunt.parser.TokenType.Companion.DERIVING
import komu.blunt.parser.TokenType.Companion.END
import komu.blunt.parser.TokenType.Companion.OR
import komu.blunt.parser.TokenType.Companion.TYPE_OR_CTOR_NAME
import komu.blunt.types.ConstructorDefinition
import komu.blunt.types.Qualified
import komu.blunt.types.Type
import komu.blunt.types.quantify
import java.util.*

internal class DataTypeParser(val lexer: Lexer, val typeParser: TypeParser) {

    // data <type> <var>* = <constructor>+
    fun parseDataDefinition(): ASTDefinition.Data {
        lexer.expectIndentStartToken(DATA)

        val builder = DataTypeBuilder(lexer.readTokenValue(TYPE_OR_CTOR_NAME))

        while (!lexer.readMatchingToken(ASSIGN))
            builder.addVariable(typeParser.parseTypeVariable())

        do {
            parseConstructor(builder)
        } while (lexer.readMatchingToken(OR))

        if (lexer.readMatchingToken(DERIVING)) {
            lexer.inParens {
                do {
                    builder.addAutomaticallyDerivedClass(lexer.readTokenValue(TYPE_OR_CTOR_NAME))
                } while (lexer.readMatchingToken(COMMA))
            }
        }

        lexer.expectToken(END)

        return builder.build()
    }

    private fun parseConstructor(builder: DataTypeBuilder): Unit {
        val constructorName = lexer.readTokenValue(TYPE_OR_CTOR_NAME)

        val args = ArrayList<Type>()
        while (!lexer.nextTokenIs(OR) && !lexer.nextTokenIs(END) && !lexer.nextTokenIs(DERIVING))
            args.add(typeParser.parseTypePrimitive())

        builder.addConstructor(constructorName, args)
    }

    private class DataTypeBuilder(val typeName: String) {

        private val vars = ArrayList<Type.Var>()
        private val constructors = ArrayList<ConstructorDefinition>()
        private val derivedClasses = ArrayList<String>()
        private var constructorIndex = 0

        fun addVariable(variable: Type.Var) {
            vars.add(variable)
        }

        fun addConstructor(constructorName: String, args: List<Type>) {
            val scheme = Qualified.simple(Type.function(args, type)).quantify(vars)
            constructors.add(ConstructorDefinition(constructorIndex++, constructorName, scheme, args.size))
        }

        private val type: Type
            get() = Type.generic(typeName, vars)

        fun addAutomaticallyDerivedClass(className: String) {
            derivedClasses.add(className)
        }

        fun build(): ASTDefinition.Data =
            ASTDefinition.Data(typeName, type, constructors, derivedClasses)
    }
}
